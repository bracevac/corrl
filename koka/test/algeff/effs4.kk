// test state as fun result from a handler (where resume is captured)

effect state<s> {
  get()    : s;
  set(x:s) : ()
}

val state = handler {
  return x -> fun(s) : cps _ { x }     // needs an annotation to bring it in the cps effect
  get()    -> fun(s) { resume(s)(s) }
  set(t)   -> fun(s) { resume(())(t) }
}

fun test() {
  set(2)
  if (get() > 0) then "positive" else "negative or zero"
}

fun main() {
  uncps{ state(test)(0).println }
}