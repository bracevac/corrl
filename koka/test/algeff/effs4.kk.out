[01;30moperator branch (state) get: resume normal
[00;39;49m[00m[01;30moperator branch (state) set: resume normal
[00;39;49m[00mpositive
.Op-get : forall<a> .op-get<a>
.Op-set : forall<a> (x : a) -> .op-set<a>
.Ops-get : forall<a> (get : .op-get<a>) -> .ops-state<a>
.Ops-set : forall<a> (set : .op-set<a>) -> .ops-state<a>
.tag-state : string
get : forall<a> () -> (state<a>) a
main : () -> console ()
set : forall<a> (x : a) -> (state<a>) ()
state : forall<a,b> (action : () -> (state<a>) b) -> ((s : a) -> b)
test : () -> (state<int>) string